storage/innobase/sync/srw_lock.cc

```diff
@@ -392,14 +402,55 @@ template void ssux_lock_impl<false>::wr_wait(uint32_t);
 template<bool spinloop>
 void ssux_lock_impl<spinloop>::rd_wait()
 {
-  for (;;)
+  if (spinloop)
+  {
+    const unsigned delay= srw_pause_delay();
+
+    for (auto spin= srv_n_spin_wait_rounds; spin; spin--)
+    {
+      srw_pause(delay);
+      if (rd_lock_try())
+        return;
+    }
+  }
+
+  uint32_t writer_lock= writer.lock.load(std::memory_order_acquire);
+
+  if (writer.HOLDER & writer_lock)
   {
-    writer.wr_lock();
-    bool acquired= rd_lock_try();
-    writer.wr_unlock();
+    writer_lock= 1 + writer.lock.fetch_add(1, std::memory_order_acquire);
+
+    bool acquired;
+
+    while (writer.HOLDER & writer_lock)
+    {
+      writer.wait(writer_lock);
+      acquired= rd_lock_try();
+      if (acquired)
+        goto acquired;
+      writer_lock = writer.lock.load(std::memory_order_acquire);
+    }
```

Comment:
If `rd_lock_try()` returned `false` because the `readers` lock word value was exactly `WRITER`, we must invoke `writer.wake()` to pass the `writer.wait()` signal that we had consumed, to a concurrent `wr_lock()` that just succeeded.

Reply:
0ba2fbe07a8e5d5fb59015cd137e27ce39533814 addresses this.